js/gateio.js

```diff
@@ -2313,6 +2313,136 @@ module.exports = class gateio extends Exchange {
+        if (size > 0) {
+            side = 'buy';
+        } else if (size < 0) {
+            side = 'sell';
+        }
+        const maintenanceRate = this.safeValue (position, 'maintenance_rate');
+        const markPrice = this.safeValue (position, 'mark_price');
+        // const contractSize = this.safeValue (position, 'size');
+        const notional = this.safeValue (position, 'value');
+        const collateral = this.safeValue (position, 'margin');
+        const marginRatio = Precise.stringDiv (maintenanceRate, collateral);
+        const initialMargin = '';
+        const unrealisedPnl = this.safeValue (position, 'unrealised_pnl');
+        return {
+            'info': position,
+            'symbol': market['symbol'],
+            'timestamp': now,
+            'datetime': this.iso8601 (now),
+            'initialMargin': collateral,
+            'initialMarginPercentage': Precise.stringDiv (initialMargin, collateral),
+            'maintenanceMargin': Precise.stringMul (maintenanceRate, notional),
+            'maintenanceMarginPercentage': maintenanceRate,
+            'entryPrice': this.safeValue (position, 'entry_price'),
+            'notional': notional,
+            'leverage': this.safeValue (position, 'leverage'),
+            'unrealizedPnl': this.safeValue (position, 'unrealised_pnl'),
+            'contracts': size,
+            'contractSize': Precise.stringDiv (notional, Precise.stringMul (size, markPrice)),
+            //     realisedPnl: position['realised_pnl'],
+            'marginRatio': marginRatio,
+            'liquidationPrice': this.safeValue (position, 'liq_price'),
+            'markPrice': markPrice,
+            'collateral': this.safeValue (position, 'margin'),
+            'marginType': undefined,
+            'side': side,
+            'percentage': Precise.stringDiv (unrealisedPnl, initialMargin),
+        };
+    }
+
+    parsePositions (positions) {
+        const result = [];
+        for (let i = 0; i < positions.length; i++) {
+            result.push (this.parsePosition (positions[i]));
+        }
+        return result;
+    }
+
+    async fetchPositions (symbols = undefined, params = {}) {
+        await this.loadMarkets ();
+        const defaultType = this.safeString2 (this.options, 'fetchPositions', 'defaultType', 'futures');
+        const type = this.safeString (params, 'type', defaultType);
+        const method = this.getSupportedMapping (type, {
+            'swap': 'privateFuturesGetSettlePositions',
+            'futures': 'privateDeliveryGetSettlePositions',
+        });
+        let positions = [];
+        const settlementCurrencies = this.getSettlementCurrencies (type, 'fetchPositions');
+        for (let c = 0; c < settlementCurrencies.length; c++) {
+            const request = {
+                'settle': settlementCurrencies[c],
+            };
+            const response = await this[method] (request);
+            positions = positions.concat (response);
```

Comment:
need to do, this.arrayConcat here

Reply:
actually you should make multiple calls in one ccxt method

Reply:
> actually you should make multiple calls in one ccxt method

I guess you meant _"should **not**"_.

Reply:
So then do we require that the settle currency be passed in via params? And only return the positions in the matching settle currency?

Reply:
> So then do we require that the settle currency be passed in via params? And only return the positions in the matching settle currency?

Yes, if the exchange requires a symbol – ccxt has to require it from the user as well.

Reply:
Would it be better to have the `settleId` be optional? And then make calls to fetch both the USDT and BTC settled contracts if it's not provided?

Reply:
@samgermain the rate limiting rule across the lib is "one http request per unified method call" – we don't really emulate those loops (the user can do that on his side, if necessary). We only do more than one request in a single method if we absolutely have to (this is very rare, in most cases we don't have to).

Reply:
ok i changed this, I'm getting a bug that I'll take a better look at later